File: backend/app/schemas/geo.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, field_serializer

MapScope = Literal["me", "mates"]


class VenueDrinker(BaseModel):
    username: str
    entry_count: int
    liters: float


class VenueMapPoint(BaseModel):
    bar_id: int
    name: str
    latitude: float
    longitude: float
    address: str | None
    postcode: str | None
    is_closed: bool
    entry_count: int
    total_liters: float
    last_visit: datetime
    drinkers: tuple[VenueDrinker, ...]  # sorted by liters desc

    # JSON only: tag the naive-UTC value with +00:00 so browsers convert to local time
    @field_serializer("last_visit", when_used="json")
    def serialize_as_utc(self, v: datetime) -> datetime:
        return v.replace(tzinfo=UTC)


class VenueMapResponse(BaseModel):
    scope: MapScope
    venues: tuple[VenueMapPoint, ...]

File: frontend/src/types/geo.ts
export type MapScope = "me" | "mates";

export interface VenueDrinker {
  username: string;
  entry_count: number;
  liters: number;
}

export interface VenueMapPoint {
  bar_id: number;
  name: string;
  latitude: number;
  longitude: number;
  address: string | null;
  postcode: string | null;
  is_closed: boolean;
  entry_count: number;
  total_liters: number;
  last_visit: string;
  drinkers: VenueDrinker[];
}

export interface VenueMapResponse {
  scope: MapScope;
  venues: VenueMapPoint[];
}

export interface ArrondissementStat {
  arrondissement: number; // 1-20
  entry_count: number;
  total_liters: number;
}

export interface ArrondissementMapResponse {
  scope: MapScope;
  arrondissements: ArrondissementStat[];
}
